### PR TITLE
Fix consistency issue with sparkjava demo tests.

### PR DESCRIPTION
**Pull request by ajkannan**
Consistency issues were likely involved with #85.

/cc @aozarov mind taking a look to check that consistency bugs are now fixed?

### CODE REVIEW

File: managed_vms/sparkjava/src/main/java/com/google/appengine/sparkdemo/UserService.java [hunk 1]
```diff
@@ -83,7 +89,9 @@ public User createUser(String name, String email) {
    */
   public String deleteUser(String id) {
     Key key = keyFactory.newKey(id);
-    datastore.delete(key);
+    if (datastore.get(key) != null) {
```
Comment (aozarov) [inline managed_vms/sparkjava/src/main/java/com/google/appengine/sparkdemo/UserService.java]: why not just delete?

**Comment by aozarov** [comment]
Change Looks good.

File: managed_vms/sparkjava/src/main/java/com/google/appengine/sparkdemo/UserService.java (hunk 1 above)
Comment (ajkannan) [inline managed_vms/sparkjava/src/main/java/com/google/appengine/sparkdemo/UserService.java]: Fixed.

**Comment by tswast** [comment]
I agree that it does seem most likely that this flakiness is due to eventual consistency issues.

I don't think this is a good recommended practice (to make all user entities share a parent entity). This introduces a global lock on all user reads/writes. This is definitely not good for datastore. One would be better off using a SQL database than writing Datastore code like this. (We are limited to about 1 write to the whole users table per second with this change. https://cloud.google.com/datastore/docs/concepts/structuring_for_strong_consistency -- last paragraph "no more than 1 write per second (the supported limit for entity groups)")

There a couple ways around the eventual consistency.

1. In a standard runtime App Engine app, the simplest would be to allow injecting in a local Datastore service (which has options for turning off eventual consistency). See the guide for unit testing App Engine services. https://cloud.google.com/appengine/docs/java/tools/localunittesting It looks like a similar thing is possible by starting a local development server and connecting to that when using the gcloud APIs. https://cloud.google.com/datastore/docs/tools/devserver There is a `--consistency=` option, which we can set to 1.0.
2. Remove uses of `getAllUsers()` in the tests. For example: the failure in #85 for `testDeleteUser` could have been avoided completely by doing a `datastore.get` instead of a `query` to verify that the user is deleted. If we do want to use `getAllUsers` in a test, we should not be looking for a specific number of results, only verify that the command succeeded.

Option 2 is probably showing users the best example of how to use and test Datastore, and manage eventual consistency. Option 1 is okay if easy to do, but it does completely ignore eventual consistency, which may be bad for users who try to follow this sample.

**Comment by waprin** [comment]
Tim's recommendations make sense for the simpler tests that you do expect to be non-flaky.But instead of replacing the tests with non-flaky ones, consider just adding the non-flaky ones and marking the flaky one as such to be only run in certain contexts and possibly to be retried a few times. We are being strongly encouraged to run good e2e tests, so some flaky tests are inevitable, so sooner or later you will probably need a broader strategy to accommodate flakiness.

**Comment by ajkannan** [comment]
Since it's a relatively simple test, I pushed another commit using option 2 as Tim suggested.

**Comment by tswast** [comment]
LGTM. Thank you!

**Pull request merged**